worker/instancepoller/worker.go

```diff
@@ -85,57 +180,320 @@ func (u *updaterWorker) Wait() error {
-			CredentialAPI: u.config.CredentialAPI,
-		},
-	)
+func (u *updaterWorker) loop() error {
+	watcher, err := u.config.Facade.WatchModelMachines()
 	if err != nil {
 		return errors.Trace(err)
 	}
-	if err := u.catacomb.Add(u.aggregator); err != nil {
+	if err := u.catacomb.Add(watcher); err != nil {
 		return errors.Trace(err)
 	}
-	watcher, err := u.config.Facade.WatchModelMachines()
+
+	shortPollTimer := u.config.Clock.NewTimer(ShortPoll)
+	longPollTimer := u.config.Clock.NewTimer(LongPoll)
+	defer func() {
+		_ = shortPollTimer.Stop()
+		_ = longPollTimer.Stop()
+	}()
+
+	for {
+		select {
+		case <-u.catacomb.Dying():
+			return u.catacomb.ErrDying()
+		case ids, ok := <-watcher.Changes():
+			if !ok {
+				return errors.New("machines watcher closed")
+			}
+
+			for i := range ids {
+				tag := names.NewMachineTag(ids[i])
+				if err := u.queueMachineForPolling(tag); err != nil {
+					return err
+				}
+			}
+		case <-shortPollTimer.Chan():
+			if err := u.pollGroupMembers(shortPollGroup); err != nil {
+				return err
+			}
+			shortPollTimer.Reset(ShortPoll)
+		case <-longPollTimer.Chan():
+			if err := u.pollGroupMembers(longPollGroup); err != nil {
+				return err
+			}
+			longPollTimer.Reset(LongPoll)
+		}
+
+		if u.loopCompletedHook != nil {
+			u.loopCompletedHook()
+		}
+	}
+}
+
+func (u *updaterWorker) queueMachineForPolling(tag names.MachineTag) error {
+	// If we are already polling this machine, check whether it is still alive
+	// and remove it from its poll group
+	if entry, groupType := u.lookupPolledMachine(tag); entry != nil {
+		if err := entry.m.Refresh(); err != nil {
+			return errors.Trace(err)
+		}
+		if entry.m.Life() == params.Dead {
+			u.config.Logger.Debugf("removing dead machine %q (instance ID %q)", entry.m, entry.instanceID)
+			delete(u.pollGroup[groupType], tag)
+			delete(u.instanceIDToGroupEntry, entry.instanceID)
+			return nil
+		}
+
+		// Something has changed with the machine state. Reset short
+		// poll interval for the machine and move it to the short poll
+		// group (if not already there) so we immediately poll its
+		// status at the next interval.
+		entry.resetShortPollInterval(u.config.Clock)
+		if groupType == longPollGroup {
+			delete(u.pollGroup[longPollGroup], tag)
+			u.pollGroup[shortPollGroup][tag] = entry
+			u.config.Logger.Debugf("moving machine %q (instance ID %q) to long poll group", entry.m, entry.instanceID)
+		}
+		return nil
+	}
+
+	// Get information about the machine
+	m, err := u.config.Facade.Machine(tag)
 	if err != nil {
 		return errors.Trace(err)
 	}
-	if err := u.catacomb.Add(watcher); err != nil {
+
+	// We don't poll manual machines, instead we're setting the status to 'running'
+	// as we don't have any better information from the provider, see lp:1678981
+	isManual, err := m.IsManual()
+	if err != nil {
 		return errors.Trace(err)
 	}
-	return watchMachinesLoop(u, watcher, u.config.Logger)
+
+	if isManual {
+		machineStatus, err := m.InstanceStatus()
```

Comment:
The [original](https://github.com/juju/juju/blob/develop/worker/instancepoller/updater.go#L144-L149) implementation uses `m.Status` which I think is incorrect. AFAICT from the documentation in the `status` package , `Status` returns the agent's status and would normally be `Created` whereas the instance status can actually be `Running`.

I have changed the logic in this implementation to use `InstanceStatus`.